src/workerd/api/tracing-module.c++

```diff
@@ -0,0 +1,225 @@
+      s->end();
+    }
+  }
+}
+
+void InternalSpan::setTag(jsg::Lock& js, const kj::String key, const kj::Maybe<kj::OneOf<bool, double, kj::String>> maybeValue) {
+  kj::OneOf<bool, double, kj::String> tagValue;
+
+  KJ_IF_SOME(value, maybeValue) {
+    KJ_SWITCH_ONEOF(value) {
+      KJ_CASE_ONEOF(s, kj::String) {
+        tagValue = kj::str(s);
+      }
+      KJ_CASE_ONEOF(b, bool) {
+        tagValue = b;
+      }
+      KJ_CASE_ONEOF(d, double) {
+        tagValue = d;
+      }
+    }
+
+    KJ_SWITCH_ONEOF(impl) {
+      KJ_CASE_ONEOF(s, kj::Own<InternalSpanImpl>) {
+        s->setTag(kj::ConstString(kj::str(key)), kj::mv(tagValue));
+      }
+      KJ_CASE_ONEOF(s, IoOwn<InternalSpanImpl>) {
+        s->setTag(kj::ConstString(kj::str(key)), kj::mv(tagValue));
+      }
+    }
+  }
+}
+
+bool InternalSpan::getIsRecording() {
+  KJ_SWITCH_ONEOF(impl) {
+    KJ_CASE_ONEOF(s, kj::Own<InternalSpanImpl>) {
+      return s->getIsRecording();
+    }
+    KJ_CASE_ONEOF(s, IoOwn<InternalSpanImpl>) {
+      return s->getIsRecording();
+    }
+  }
+  KJ_UNREACHABLE;
+}
+
+// InternalSpanImpl implementation
+InternalSpanImpl::InternalSpanImpl(SpanBuilder span): span(kj::mv(span)) {}
+
+InternalSpanImpl::InternalSpanImpl(decltype(nullptr)): span(kj::none) {}
+
+InternalSpanImpl::~InternalSpanImpl() noexcept(false) {
+  end();
+}
+
+void InternalSpanImpl::end() {
+  KJ_IF_SOME(s, span) {
+    s.end();
+    span = kj::none;
+  }
+}
+
+void InternalSpanImpl::setTag(kj::ConstString key, kj::OneOf<bool, double, kj::String> value) {
+  KJ_IF_SOME(s, span) {
+    KJ_SWITCH_ONEOF(value) {
+      KJ_CASE_ONEOF(b, bool) {
+        s.setTag(kj::mv(key), b);
+      }
+      KJ_CASE_ONEOF(d, double) {
+        s.setTag(kj::mv(key), d);
+      }
+      KJ_CASE_ONEOF(str, kj::String) {
+        s.setTag(kj::mv(key), kj::mv(str));
+      }
+    }
+  }
+}
+
+bool InternalSpanImpl::getIsRecording() {
+  return span != kj::none;
+}
+
+SpanParent InternalSpanImpl::makeSpanParent() {
+  KJ_IF_SOME(s, span) {
+    return SpanParent(s);
+  } else {
+    return SpanParent(nullptr);
+  }
+}
+
+SpanParent InternalSpan::makeSpanParent() {
+  KJ_SWITCH_ONEOF(impl) {
+    KJ_CASE_ONEOF(s, kj::Own<InternalSpanImpl>) {
+      return s->makeSpanParent();
+    }
+    KJ_CASE_ONEOF(s, IoOwn<InternalSpanImpl>) {
+      return s->makeSpanParent();
+    }
+  }
+  KJ_UNREACHABLE;
+}
+
+jsg::Ref<InternalSpan> TracingModule::startSpan(jsg::Lock& js, const kj::String name) {
+  auto spanName = kj::ConstString(kj::str(name));
+
+  if (IoContext::hasCurrent()) {
+    auto& ioContext = IoContext::current();
+    auto span = ioContext.makeUserTraceSpan(kj::mv(spanName));
+    auto impl = kj::heap<InternalSpanImpl>(kj::mv(span));
+    return js.alloc<InternalSpan>(ioContext.addObject(kj::mv(impl)));
+  } else {
+    // When no IoContext is available, create a null span
+    auto impl = kj::heap<InternalSpanImpl>(nullptr);
+    return js.alloc<InternalSpan>(kj::mv(impl));
```

Comment:
Rather than having `InternalSpan` have an `kj::OneOf<kj::Own<T>, IoOwn<T>>`, when one of the cases is always going to be a null span, perhaps it's better for it to hold a `kj::Maybe<IoOwn<T>>` such that it is completely non-op when there is no `IoContext`